ste/xferLifetimeTracker.go

```diff
@@ -0,0 +1,300 @@
+		newAtomicInt := func() *int64 {
+			var n int64
+			return &n
+		}
+
+		_requestLifetimeTrackerSingleton = &RequestLifetimeTracker{
+			requestActionQueue:         make(chan requestAction, 600), // 2x our max number of threads, that way the queue should never block up.
+			liveRequestInitiationTimes: make(map[uuid.UUID]time.Time),
+			requestLifetimeBuckets:     &requestLifetimeBucket{},
+			bucketSwapTicker:           time.NewTicker(requestBucketLifetime),
+			bucketSwapCond:             sync.NewCond(&sync.Mutex{}),
+
+			atomicSimultaneousLiveRequestCount: newAtomicInt(),
+			atomicUnexpectedRequestLifetime:    newAtomicInt(),
+
+			atomicRequestLifetimeSum:     newAtomicInt(),
+			atomicRequestCompletionCount: newAtomicInt(),
+			atomicRequestMaxLifetime:     newAtomicInt(),
+			atomicRequestAvgLifetime:     newAtomicInt(),
+
+			atomicRunningRequestAverageLifetime: newAtomicInt(),
+			atomicRunningRequestCount:           newAtomicInt(),
+			atomicRunningRequestMaxLifetime:     newAtomicInt(),
+		}
+
+		_requestLifetimeTrackerSingleton.requestLifetimeBuckets.Insert(&common.LinkedList[time.Duration]{})
+
+		go _requestLifetimeTrackerSingleton.queueWorker()
+	})
+
+	return _requestLifetimeTrackerSingleton
+}
+
+type requestLifetimeBucket = common.LinkedList[*common.LinkedList[time.Duration]]
+type requestAction struct {
+	ID     uuid.UUID
+	Action uint
+
+	SubmitTime time.Time
+}
+
+type RequestLifetimeTracker struct {
+	atomicSimultaneousLiveRequestCount *int64 // how many requests are currently live
+	atomicUnexpectedRequestLifetime    *int64 // Are our longest-living requests alive too long?
+
+	// Keeping track of a total average, for debug purposes.
+	atomicRequestLifetimeSum     *int64 // the sum of all requests' lifetime
+	atomicRequestMaxLifetime     *int64 // The highest lifetime we've *ever* seen.
+	atomicRequestCompletionCount *int64 // how many requests have been completed
+	atomicRequestAvgLifetime     *int64 // the average of all requests' lifetime
+
+	// This is the useful bit. Keep track of in-flight requests and recent historical data.
+	requestActionQueue         chan requestAction
+	liveRequestInitiationTimes map[uuid.UUID]time.Time
+	requestLifetimeBuckets     *requestLifetimeBucket // Divided into n buckets of duration
+	bucketSwapTicker           *time.Ticker
+	bucketSwapCond             *sync.Cond
+
+	// more atomic values for read-only
+	atomicRunningRequestAverageLifetime *int64
+	atomicRunningRequestMaxLifetime     *int64
+	atomicRunningRequestCount           *int64
+}
+
+// workers sit in the background and manage the running averages and buckets, ticking with the bucket ticker.
+
+func (r *RequestLifetimeTracker) queueWorker() {
+	requestBucketCount := int64(requestTrackingWindow / requestBucketLifetime)
+
+	for {
+		select { // This thread should die with the program to prevent any hangups
+		case <-r.bucketSwapTicker.C:
+			var rCount, lCount int64
+			var rSum, rAvg, rMax time.Duration
```

Comment:
It might be good to comment what these are tracking - Are the r prefixed ones, historical request metrica and the l prefixed ones live request ones?